persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -360,20 +364,31 @@ openSimpleConn = openSimpleConnWithVersion getServerVersion
 -- @since 2.9.1
 openSimpleConnWithVersion :: (PG.Connection -> IO (Maybe Double)) -> LogFunc -> PG.Connection -> IO SqlBackend
 openSimpleConnWithVersion getVerDouble logFunc conn = do
-    smap <- newIORef $ Map.empty
+    smap <- mkSimpleStatementCache
     serverVersion <- oldGetVersionToNew getVerDouble conn
-    return $ createBackend logFunc serverVersion smap conn
+    return $ createBackend logFunc serverVersion (mkStatementCache smap) conn
+
+underlyingConnectionKey :: Vault.Key PG.Connection
+underlyingConnectionKey = unsafePerformIO Vault.newKey
+{-# NOINLINE underlyingConnectionKey #-}
+
+-- | Access underlying connection, returning 'Nothing' is the 'SqlBackend'
+-- provided isn't backed by postgresql-simple.
+--
+-- @since 2.14.0
+getSimpleConn :: (BackendCompatible SqlBackend backend) => backend -> Maybe PG.Connection
+getSimpleConn = Vault.lookup underlyingConnectionKey <$> getConnVault
 
 -- | Create the backend given a logging function, server version, mutable statement cell,
 -- and connection.
 createBackend :: LogFunc -> NonEmpty Word
-              -> IORef (Map.Map Text Statement) -> PG.Connection -> SqlBackend
+              -> StatementCache -> PG.Connection -> SqlBackend
 createBackend logFunc serverVersion smap conn =
     maybe id setConnPutManySql (upsertFunction putManySql serverVersion) $
     maybe id setConnUpsertSql (upsertFunction upsertSql' serverVersion) $
     setConnInsertManySql insertManySql' $
     maybe id setConnRepsertManySql (upsertFunction repsertManySql serverVersion) $
-    mkSqlBackend MkSqlBackendArgs
+    modifyConnVault (Vault.insert underlyingConnectionKey conn) $ mkSqlBackend MkSqlBackendArgs
```

Comment:
To get access to the underlying backend, use the `RawPostgresql` type https://github.com/yesodweb/persistent/blob/69bdefd3978aa27004fb118f301aae0fb1c40640/persistent-postgresql/Database/Persist/Postgresql.hs#L1845

Reply:
This isn't sufficient for instrumentation of existing codebases unfortunately. The value lets us wrap existing connections without requiring pervasive downstream changes.